### ISSUE TITLE
Autodetect type by assignment if possible

### ISSUE THREAD

**Issue by MaxGraey** [dcodeIO/AssemblyScript#41]
Instead this
```ts
let array: Array< float > = new Array< float >(1);
```
will be great write simpler:
```ts
let array = new Array< float >(1);
```

The same for plain literal types:
```ts
var _num = 1.0; // type = float64
var _int = 1; // type = int32
var _boolean = true // type = bool
...
```

What you thinking about this? How much effort need to make this working?

**Comment by dcodeIO** [comment]
Detecting types implicitly could become a bit problematic, especially when wasm32/64 is both feasible. Haven't decided anything yet, still have to see how things play out once people start using them, but once thing we could do already would allow omitting the type parameter on `Array<int>` if the left hand side already defines the type as `Array<int>`, like:

```ts
let array: int[] = new Array(10);
```

Thoughts?

**Comment by MaxGraey** [comment]
Hmm, in many languages implicitly typing resolved, for example [swift](https://www.tutorialspoint.com/swift/swift_literals.htm) and [C#](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/implicitly-typed-local-variables). Both have got 32/64 types. But I agree typescript has some gap to figure out how interpret type implicitly. I think it can be resolved by special compiler option. Thoughts?

**Comment by dcodeIO** [comment]
One issue I see there is that other languages have special postfixes to mark literals to be of a specific type. C# for example has the `L` postfix to mark a numeric literal as a long type. TypeScript doesn't have these, and that bothers me because implicit types would be limited to just some types, and that doesn't feel right to me.

**Comment by MaxGraey** [comment]
How about this?

```ts
let longType = 123 as int32;
let floatType = 1e-5 as double;
```

**Comment by dcodeIO** [comment]
Yeah, that would work, but these are actually more characters to type than just specifying the type.

**Comment by MaxGraey** [comment]
I agree with you. Without short literal suffixes this no make sense.

**Comment by dcodeIO** [comment]
One alternative here is using type assertion expressions for that, like `let longType = <long>123;`. Haven't yet though this through fully.

**Comment by MaxGraey** [comment]
What about about implicit typing when assignment from one variable to another variable?

```ts
let a: float = 2.0;

let b = a;
```

**Comment by dcodeIO** [comment]
This might require specific code at this time, because, currently, the contextual type is always determined by evaluating the lhs expression.

**Comment by MaxGraey** [comment]
Ok, this not very important feature. May in some day
